docs/server/code.md

```diff
@@ -0,0 +1,135 @@
+
+For example our pkg/mobile/integration has business logic around integrating services with each other. However it knows nothing about kubernetes, instead it relies on implementations of 
+its required core interfaces in order to fulfill its business logic.
+
+#### External Layers (tests, web, cli, external services)
+
+The remaining packages make up our concrete implementations of the core interfaces required by the Business logic layer. They can depend on any of the internal layers
+they can know everything about a specific database or external service. They should not leak this specific knowledge.
+A special note on tests. Even though they are within the same package as the code they are testing, they are an external interface acting on the code they are testing. As such  they often are part of the special
+```_test``` package which is treated as a separate package and it only has access to the external API of the package it is testing.
+  
+
+
+```
+pkg
+├── cli #cli interface (external layer)
+│   └── cmd #cli based parsing and handling.
+├── clients #client builders for external services (external layer) 
+├── data # concrete implementations of the mobile/interfaces.go for data manipulation
+├── k8s  # kubernetes client and specific logs (external layer) 
+├── mobile
+│   ├── interfaces.go # domain logic interfaces (AppRepo, ServiceRepo etc)
+│   ├── types.go # domain models (mobile app, mobile services, mobile config)
+│   ├── app # core app based business logic
+│   ├── integration # core integration business logic
+│   └── sdk # core sdk based logic
+├── mock # external for mocking in tests
+├── openshift # external specific to openshift
+└── web # external http handlers and parsers (only resposible for accepting the request parsing and handing off to business logic)
+    ├── headers
+    └── middleware
+
+```
+
+
+### How Authentication and Authorization works
+
+**Authentication**
+
+The ui for the MCP is embedded into the OpenShift UI, we are able to reuse the bearer token created when the user first logs into OpenShift.
+From the UI, this token is sent with each request to the server. Using this token the access middleware first checks if it needs to check access to the currently
+requested URL, if it does, it makes a request to the user read endpoint in OpenShift. If we get a valid response, then we know the user is a 
+valid user for this instance of OpenShift and we can proceed.
+
+**Authorization**
+
+Once a user is authenticated, we use their bearer token in order to interact with the kubernetes and openshift API. In this way we offload 
+authorisation to the cluster. However is some places it is useful to check access to a particular resource within a given namespace. To do this we 
+do a localaccessreview. 
```

Comment:
not sure if it matters, but I've actually ended up using a `localsubjectaccessreviews`

Reply:
any docs or code we can link to that gives more info on what this is?